File: HelloWorld/Cpp/dll/HelloWorldDataItem.cpp
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

#include "StdAfx.h"
#include "HelloWorldDataItem.h"

// Returns the instance of CHelloWorldDataItem associated with the input DkmStackContext
// object. If there is not currently an associated CHelloWorldDataItem, a new data item
// will be created.
HRESULT CHelloWorldDataItem::GetInstance(
    DkmStackContext* pContext,
    CHelloWorldDataItem** ppStateObject
    )
{
    HRESULT hr;

    // If there is already an associated item, return it.
    hr = GetExistingInstance(pContext, ppStateObject);
    if (hr == S_OK)
        return hr;

    // Otherwise create a new object
    CComObject<CHelloWorldDataItem>* pComObject;
    hr = CComObject<CHelloWorldDataItem>::CreateInstance(&pComObject);
    if (FAILED(hr))
    {
        return hr;
    }

    // Assign it to a CComPtr so that it is AddRef'ed
    CComPtr<CHelloWorldDataItem> pCreatedInstance(pComObject);

    // Then associate the new data item with pContext
    hr = pContext->SetDataItem(DkmDataCreationDisposition::CreateNew, pCreatedInstance);
    if (FAILED(hr))
    {
        // NOTE: Since call stack walking is already synchronized, this should never 
        // fail (absent out-of-memory). In other scenarios, it might be important to 
        // handle the case that a component is trying to set the data item 
        // simultaneously from multiple threads.
        return hr;
    }

    // Return the new object
    *ppStateObject = pCreatedInstance.Detach();
    return S_OK;
}

// Returns the instance of CHelloWorldDataItem associated with the input DkmStackContext
// object. If there is not currently an associated CHelloWorldDataItem, this method will
// fail.
HRESULT CHelloWorldDataItem::GetExistingInstance(
    DkmStackContext* pContext,
    CHelloWorldDataItem** ppStateObject
    )
{
    return pContext->GetDataItem(ppStateObject);
}

File: HelloWorld/Cs/dll/HelloWorldDataItem.cs
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using Microsoft.VisualStudio.Debugger;
using Microsoft.VisualStudio.Debugger.CallStack;

namespace HelloWorld
{
    /// <summary>
    /// Defines the two possible states the HelloWorld stack frame filter can be in.
    /// </summary>
    internal enum State
    {
        /// <summary>
        /// The initial state (state that we start out in). At this point, we
        /// haven't seen any stack frames.
        /// </summary>
        Initial,

        /// <summary>
        /// The state that we are in after we added the '[Hello World]' frame.
        /// </summary>
        HelloWorldFrameAdded
    };

    /// <summary>
    /// HelloWorldDataItem is an internal object used to hold the data which the HelloWorld 
    /// component associates with a DkmStackContext. In other words, this is a state-store which
    /// the hello world sample can use to hold data associated with a stack walk session.
    /// </summary>
    internal class HelloWordDataItem : DkmDataItem
    {
        // Object is created from GetInstance
        private HelloWordDataItem()
        {
        }

        public State State
        {
            get;
            set;
        }

        /// <summary>
        /// Returns the instance of HelloWorldDataItem associated with the input DkmStackContext
        /// object. If there is not currently an associated HelloWorldDataItem, a new data item
        /// will be created.
        /// </summary>
        /// <param name="context">Object to obtain the data item from</param>
        /// <returns>The associated data item</returns>
        public static HelloWordDataItem GetInstance(DkmStackContext context)
        {
            HelloWordDataItem item = GetExistingInstance(context);
            if (item != null)
                return item;

            item = new HelloWordDataItem();
            context.SetDataItem<HelloWordDataItem>(DkmDataCreationDisposition.CreateNew, item);

            return item;
        }

        /// <summary>
        /// Returns the instance of HelloWorldDataItem associated with the input DkmStackContext
        /// object. If there is not currently an associated HelloWorldDataItem, return null.
        /// </summary>
        /// <param name="context">Object to obtain the data item from</param>
        /// <returns>[OPTIONAL] The associated data item</returns>
        public static HelloWordDataItem GetExistingInstance(DkmStackContext context)
        {
            return context.GetDataItem<HelloWordDataItem>();
        }
    }
}
